sky/shell/android/update_service_android.cc

```diff
@@ -0,0 +1,129 @@
+// Copyright 2015 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "base/android/jni_string.h"
+#include "base/bind.h"
+#include "base/files/file_util.h"
+#include "base/files/file_util.h"
+#include "base/logging.h"
+#include "base/threading/worker_pool.h"
+#include "jni/UpdateService_jni.h"
+#include "mojo/data_pipe_utils/data_pipe_utils.h"
+#include "mojo/public/cpp/application/connect.h"
+#include "mojo/public/interfaces/application/service_provider.mojom.h"
+#include "mojo/services/network/public/interfaces/network_service.mojom.h"
+#include "sky/shell/shell.h"
+
+namespace sky {
+namespace shell {
+
+namespace {
+
+// This helper exists to keep the URLLoaderPtr alive until we call the callback.
+// I'm not sure if this is necessary, though.
+static void CallResponseCallback(
+    mojo::URLLoaderPtr loader,
+    base::Callback<void(mojo::URLResponsePtr)> callback,
+    mojo::URLResponsePtr response) {
+  callback.Run(response.Pass());
+}
+
+static void FetchURL(mojo::NetworkService* network_service,
+                     const std::string& url,
+                     base::Callback<void(mojo::URLResponsePtr)> callback) {
+  mojo::URLLoaderPtr loader;
+  network_service->CreateURLLoader(GetProxy(&loader));
+
+  mojo::URLRequestPtr request = mojo::URLRequest::New();
+  request->url = url;
+  request->auto_follow_redirects = true;
+
+  mojo::URLLoader* raw_loader = loader.get();
+  raw_loader->Start(
+      request.Pass(),
+      base::Bind(&CallResponseCallback, base::Passed(&loader), callback));
+}
+
+class UpdateTask {
+ public:
+  UpdateTask(JNIEnv* env, jobject update_service, std::string data_dir)
+      : final_path_(base::FilePath(data_dir).AppendASCII("app.skyx")) {
+    update_service_.Reset(env, update_service);
+
+    mojo::ServiceProviderPtr service_provider =
+        CreateServiceProvider(Shell::Shared().service_provider_context());
+    mojo::ConnectToService(service_provider.get(), &network_service_);
+  }
+
+  void DownloadSkyx(const std::string& url) {
+    LOG(INFO) << "Update downloading to: " << final_path_.value();
+    if (!base::CreateTemporaryFile(&temp_path_)) {
+      CallOnFinished();
+      return;
+    }
+
+    FetchURL(network_service_.get(), url,
+             base::Bind(&UpdateTask::OnResponse, base::Unretained(this)));
```

Comment:
Maybe it would make more sense if we made the UpdateTask explicitly owned in Java by the UpdateService.  Then the update service could have the ability to cancel the update or otherwise reclaim the resources being consumed by the UpdateTask,